client/Cargo.toml

```diff
@@ -34,12 +34,16 @@ log = { workspace = true }
 thiserror = { workspace = true }
 
 # Light Protocol
+account-compression = { workspace = true }
+light-compressed-token = { workspace = true }
```

Comment:
pls remove these dependencies we would like to keep `light-client`, `light-sdk` etc free of anchor hence program dependencies.

Reply:
seeing that you are importing struct definitions from the program dependencies.
`light-sdk` contains many copy pasted struct definitions already to avoid program dependencies. You can probably import most from there. In case some are missing fee free to add them in `light-client` as well for example for `PublicTransactionEvent` although I am not sure that you need that outside of the `test-indexer`.

Reply:
> else copy pasta struct definition where you need it

I don't like this approach tbh, its error-prone and hard to handle, especially in `test-utils` and `test-programs`, where every second test uses structs from programs and `light-sdk` / `client` simultaneously, leading to nasty code like this https://github.com/Lightprotocol/light-protocol/blob/sergey/refactor_test_indexer/test-programs/sdk-test-program/programs/sdk-test/tests/test.rs#L99-L129

wouldn't it be better to extract these common structs into separate crate, and use this crate in `*-programs` and in `client`& `sdk`?

Reply:
**I see 4 options here the tldr see for details below:**
1. upgrade anchor and extract programs, don't like it a lot of effort and will remove anchor next year
2. introduce an sdk feature to programs and extract the sdk with and without anchor
3. extract structs into one crate but just copy paste don't use this in programs
4. copy pasta as Michal did it for the sdk

leaning towards 3 or 4, @sergeytimoshin wdyt?

**Details:**
1. The reason why we haven't extracted the types so far is that 0.29.0 idl generation doesn't work anymore and bumping to anchor 0.30.0 has proven to be a nightmare. I don't think this is a solution, I want to ditch anchor completely next year to optimize our core programs.

2. Maybe an easier way is to make anchor an optional dependency for programs and create a feature which only compiles the system-program/src/sdk/* then we can import types from programs without the anchor dependency.
- add feature sdk
- make anchor optional
- import programs to sdk without default features
This could be a fast and non-intrusive way but it will increase compile time of the sdk with anchor a lot which I don't like.

3. Different way, create one crate which contains all the relevant extracted structs with copy pasta and use it. Don't import programs  Still copy pasta but everything is in one place.
    - easier to maintain than copy pasta in multiple places
    - doesn't increase compile time of the sdk

4. Or just stick to copy pasta for now, the system program instruction bindings are not going to change ever with it the structs to generate instruction data. We are just going to add new instructions going forward and possibly deprecate old once.

Reply:
Hm.
I like both approaches 3 (separate crate) and 4 (copy-paste), but there's still the problem of manual type conversions in tests like:
```
let mut sdk_compressed_account_data: Option<CompressedAccountData> = None;
if let Some(compressed_account_data) = &compressed_account.compressed_account.data {
    sdk_compressed_account_data = Some(CompressedAccountData {
        discriminator: compressed_account_data.discriminator,
        // ... lots of manual field mapping
    });
}
```

Instead of choosing between 3 and 4, I think about hybrid approach:

1. Keep copy-pasting structs in the SDK (approach 4)
2. Add conversion traits ONLY in test code using cfg(test), allowing ergonomic conversions like:
`let sdk_account: sdk::CompressedAccountWithMerkleContext = program_account.into();`

This gives us:
1. Clean SDK without program dependencies
2. A lot less pain wrt type conversion in tests
3. Still easy path to remove Anchor later

Reply:
sounds good but please make anchor optional from the start.
Michal found an easy way see:
https://github.com/Lightprotocol/light-protocol/pull/1364/files#diff-7972129256fe1dd3b281558ba8f49d1540d7c1cafb3b2033f1335a0a83645708R1-R4

Reply:
No, we can't put conversion traits in `test-utils` because of [E0117 (Only traits defined in the current crate can be implemented for arbitrary types.)](https://doc.rust-lang.org/error_codes/E0117.html).

This it what I proposing, just one conversion as an example: https://github.com/Lightprotocol/light-protocol/pull/1376/commits/7c5c7e66ff4c6f88f30dd5072db2581bb94e3372

If you ok with it, I will do the rest conversions like this one.

Reply:
the issue is that this requires sdk to import the programs which we wanted to avoid.

Reply:
Yes, correct